pkg/pyramid/file_tracker.go

```diff
@@ -0,0 +1,50 @@
+package pyramid
+
+import (
+	"sync"
+
+	"github.com/treeverse/lakefs/pkg/pyramid/params"
+)
+
+// fileTracker tracks file open requests in TierFS to avoid race conditions with cache rejection/eviction
+type fileTracker struct {
+	refMap map[string]int
+	mu     sync.Mutex
+}
+
+type decCallback func(int)
+
+func newFileTracker() *fileTracker {
+	return &fileTracker{
+		refMap: map[string]int{},
+		mu:     sync.Mutex{},
+	}
+}
+
+func (t *fileTracker) inc(path params.RelativePath) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	ref := 1
+
+	if val, ok := t.refMap[string(path)]; ok {
+		ref = val + 1
+		t.refMap[string(path)] = ref
+	} else {
+		t.refMap[string(path)] = ref
+	}
+}
+
+func (t *fileTracker) dec(path params.RelativePath, fn decCallback) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	ref := 0
+
+	if val, ok := t.refMap[string(path)]; ok {
+		ref = val - 1
+		t.refMap[string(path)] = ref
+	}
+	if ref == -1 {
+		delete(t.refMap, string(path))
+	}
```

Comment:
1. Why is -1 the point of eviction and not 0?
2. What is the meaning of `dec` an non existing path?

Reply:
0 Doesn't mean the file was flagged for eviction - it just means that nobody is currently trying to open it.
When we trigger the eviction function it performs a decrement on the file without a prior increment - which will eventually after all the open requests completed will cause it to reach -1. This is the indication that the file was requested for deletion and only then the logic will trigger